# Alignements Athena BDPM - RxNorm - Quelques stats

In [2]:
import pandas as pd

In [20]:
import os
import sys

In [32]:
vocab_path = '../../../Documents/MAD/coagulo/vocab_tables'

In [33]:
os.listdir(vocab_path)

['CONCEPT.csv',
 'CONCEPT_ANCESTOR.csv',
 'CONCEPT_CLASS.csv',
 'CONCEPT_RELATIONSHIP.csv',
 'CONCEPT_SYNONYM.csv',
 'DOMAIN.csv',
 'DRUG_STRENGTH.csv',
 'RELATIONSHIP.csv',
 'VOCABULARY.csv']

In [34]:
# Table concept
concept_path = os.path.join(vocab_path, 'concept.csv')
concept = pd.read_csv(concept_path, sep='\t')
concept.head()

C:\Users\GaëlleRimaud\AppData\Local\Temp\ipykernel_12696\1154729115.py:3: DtypeWarning: Columns (5,6,9) have mixed types. Specify dtype option on import or set low_memory=False.
  concept = pd.read_csv(concept_path, sep='\t')


concept_id                             concept_name domain_id  \
0    45756805                     Pediatric Cardiology  Provider   
1    45756804                 Pediatric Anesthesiology  Provider   
2    45756803  Pathology-Anatomic / Pathology-Clinical  Provider   
3    45756802                    Pathology - Pediatric  Provider   
4    45756801            Pathology - Molecular Genetic  Provider   

  vocabulary_id     concept_class_id standard_concept concept_code  \
0          ABMS  Physician Specialty                S  OMOP4821938   
1          ABMS  Physician Specialty                S  OMOP4821939   
2          ABMS  Physician Specialty                S  OMOP4821940   
3          ABMS  Physician Specialty                S  OMOP4821941   
4          ABMS  Physician Specialty                S  OMOP4821942   

   valid_start_date  valid_end_date invalid_reason  
0          19700101        20991231            NaN  
1          19700101        20991231            NaN  
2          19700101        20991231            NaN  
3          19700101        20991231            NaN  
4          19700101        20991231            NaN

In [35]:
# Table concept_relationship --> Table qui contient les relations entre les codes (alignements, hiérarchiques, etc)
concetp_rel_path =  os.path.join(vocab_path, 'CONCEPT_RELATIONSHIP.csv')
concept_rel = pd.read_csv(concetp_rel_path, sep='\t')
concept_rel.head()

concept_id_1  concept_id_2      relationship_id  valid_start_date  \
0       1970397       1970398     Available as box          20220128   
1       1970401       1970304     Available as box          20220128   
2       1970450       1973231  Has quantified form          20220128   
3       1970453       1970454          Constitutes          20220128   
4      42616294      42616296             Subsumes          20161201   

   valid_end_date  invalid_reason  
0        20991231             NaN  
1        20991231             NaN  
2        20991231             NaN  
3        20991231             NaN  
4        20991231             NaN

In [36]:
# Nomenclatures uniques dans concept
concept.vocabulary_id.unique()

array(['ABMS', 'ATC', 'BDPM', 'CCAM', 'CDM', 'CIM10',
       'CMS Place of Service', 'Concept Class', 'Condition Status',
       'Condition Type', 'Cost', 'Cost Type', 'Death Type', 'Device Type',
       'Domain', 'Drug Type', 'Episode', 'Gender', 'ICD10', 'ICD10CM',
       'ICD10PCS', 'Korean Revenue Code', 'LOINC', 'Language',
       'Meas Type', 'Medicare Specialty', 'Metadata', 'None', 'Note Type',
       'OMOP Extension', 'OSM', 'Obs Period Type', 'Observation Type',
       'Plan', 'Plan Stop Reason', 'Procedure Type', 'Provider',
       'Relationship', 'RxNorm', 'RxNorm Extension', 'SNOMED', 'SOPT',
       'Sponsor', 'Type Concept', 'UB04 Point of Origin',
       'UB04 Pri Typ of Adm', 'UB04 Pt dis status', 'UB04 Typ bill',
       'UCUM', 'US Census', 'Visit', 'Visit Type', 'Vocabulary'],
      dtype=object)

### Table concept pour les codes de la BDPM

In [37]:
# Table concept pour les codes de la BDPM
bdpm = concept[concept.vocabulary_id=='BDPM'][['concept_code','concept_id','concept_name','concept_class_id']].rename(columns={'concept_code' : 'bdpm_code', 
'concept_id': 'bdpm_concept_id',
'concept_name':'bdpm_libelle',
'concept_class_id' : 'bdpm_concept_class_id'})
bdpm.head()

bdpm_code  bdpm_concept_id           bdpm_libelle bdpm_concept_class_id
6838     00001         43047359             Loperamide            Ingredient
6839     00005         43045429   Acetylsalicylic Acid            Ingredient
6840     00007         43045800           Benzoic Acid            Ingredient
6841     00008         43045866             Boric Acid            Ingredient
6842     00011         43045611  Anhydrous Citric Acid            Ingredient

In [38]:
#Nombre de valeurs uniques
bdpm.nunique()

bdpm_code                44376
bdpm_concept_id          44376
bdpm_libelle             31140
bdpm_concept_class_id        6
dtype: int64

In [53]:
# Nombre de lignes par classe
bdpm.bdpm_concept_class_id.value_counts()

Drug Product    22113
Brand Name      15053
Ingredient       4310
Supplier         1504
Device           1162
Dose Form         234
Name: bdpm_concept_class_id, dtype: int64

In [45]:
bdpm[(bdpm.bdpm_concept_class_id=='Drug Product')
     & (bdpm.bdpm_libelle.str.startswith('DOLIPRANE'))].sort_values(by='bdpm_libelle')

bdpm_code  bdpm_concept_id  \
18357   3499833         43232489   
18090   3480911         43234483   
31387   5636955         43240183   
19939   3595583         43240182   
18776   3529422         43222097   
31354   5632992         43222098   
27558   4153396         43246079   
20653   3624698         43236632   
18358   3499879         43230504   
18091   3480940         43229077   
17767   3461546         43249575   
18359   3499916         43249959   
18092   3480986         43221964   
18360   3499945         43230753   
18093   3481017         43249072   
30567   5547335         43237334   
15213   3232018         43245043   
15917   3307199         43245766   
31045   5608261         43226203   
17658   3450778         43222215   
15219   3233153         43221943   
15886   3304746         43249329   
18198   3486606         43249162   
27559   4153404         43236678   
11115   2789411         43236715   
11120   2789747         43223430   
18631   3521202         43222016   
27694   4163124         43236600   
27695   4163182         43245960   
23570   3766501         43226392   

                                            bdpm_libelle bdpm_concept_class_id  
18357  DOLIPRANE 100 mg, poudre pour solution buvable...          Drug Product  
18090             DOLIPRANE 100 mg, suppositoire sécable          Drug Product  
31387                        DOLIPRANE 1000 mg, comprimé          Drug Product  
19939                        DOLIPRANE 1000 mg, comprimé          Drug Product  
18776   DOLIPRANE 1000 mg, comprimé effervescent sécable          Drug Product  
31354   DOLIPRANE 1000 mg, comprimé effervescent sécable          Drug Product  
27558                          DOLIPRANE 1000 mg, gélule          Drug Product  
20653  DOLIPRANE 1000 mg, poudre pour solution buvabl...          Drug Product  
18358  DOLIPRANE 150 mg, poudre pour solution buvable...          Drug Product  
18091                     DOLIPRANE 150 mg, suppositoire          Drug Product  
17767  DOLIPRANE 2,4 POUR CENT SANS SUCRE, suspension...          Drug Product  
18359  DOLIPRANE 200 mg, poudre pour solution buvable...          Drug Product  
18092                     DOLIPRANE 200 mg, suppositoire          Drug Product  
18360  DOLIPRANE 300 mg, poudre pour solution buvable...          Drug Product  
18093                     DOLIPRANE 300 mg, suppositoire          Drug Product  
30567                         DOLIPRANE 500 mg, comprimé          Drug Product  
15213                         DOLIPRANE 500 mg, comprimé          Drug Product  
15917            DOLIPRANE 500 mg, comprimé effervescent          Drug Product  
31045                           DOLIPRANE 500 mg, gélule          Drug Product  
17658                           DOLIPRANE 500 mg, gélule          Drug Product  
15219  DOLIPRANE 500 mg, poudre pour solution buvable...          Drug Product  
15886            DOLIPRANE ADULTES 1000 mg, suppositoire          Drug Product  
18198   DOLIPRANE CODEINE 400 mg/20 mg, comprimé sécable          Drug Product  
27559                      DOLIPRANECAPS 1000 mg, gélule          Drug Product  
11115  DOLIPRANELIQUIZ 200 mg SANS SUCRE, suspension ...          Drug Product  
11120  DOLIPRANELIQUIZ 300 mg SANS SUCRE, suspension ...          Drug Product  
18631    DOLIPRANEORODOZ 500 mg, comprimé orodispersible          Drug Product  
27694          DOLIPRANETABS 1000 mg, comprimé pelliculé          Drug Product  
27695           DOLIPRANETABS 500 mg, comprimé pelliculé          Drug Product  
23570  DOLIPRANEVITAMINEC 500 mg/150 mg, comprimé eff...          Drug Product

### Table concept pour les codes RxNorm et RxNorm Extension

In [47]:
rxnorm = concept[(concept.vocabulary_id=='RxNorm') | (concept.vocabulary_id=='RxNorm Extension') ][['concept_code','concept_id','concept_name','concept_class_id']].rename(columns={'concept_code':'rxnorm_code','concept_name':'rxnorm_libelle','concept_id' : 'rxnorm_concept_id','concept_class_id':'rxnorm_concept_class_id'})
rxnorm.head()

rxnorm_code  rxnorm_concept_id  \
893530     1042838           40229458   
893531     1042847           40229459   
893532     1049224           40232055   
893533     1371295           43525657   
893534     1371298           43525658   

                                           rxnorm_libelle  \
893530                     benzocaine 0.18 MG/MG Oral Gel   
893531                benzocaine 0.18 MG/MG Oral Ointment   
893532     benzocaine 0.18 MG/MG Oral Ointment [Topicale]   
893533  benzocaine 180 MG/ML Mucous Membrane Topical S...   
893534  benzocaine 180 MG/ML Mucous Membrane Topical S...   

       rxnorm_concept_class_id  
893530           Clinical Drug  
893531           Clinical Drug  
893532            Branded Drug  
893533           Clinical Drug  
893534            Branded Drug

In [48]:
# Nombre de valeurs uniques
rxnorm.nunique()

rxnorm_code                2454034
rxnorm_concept_id          2454034
rxnorm_libelle             2417210
rxnorm_concept_class_id         26
dtype: int64

In [49]:
# Nombre de codes par classe
rxnorm.rxnorm_concept_class_id.value_counts()

Marketed Product        602280
Branded Drug            234654
Branded Drug Box        234649
Branded Drug Comp       202643
Branded Drug Form       164614
Clinical Drug           134955
Clinical Drug Box       116427
Quant Branded Drug      114939
Brand Name              110331
Clinical Drug Comp      109411
Quant Branded Box       100410
Quant Clinical Drug      89767
Quant Clinical Box       79806
Clinical Drug Form       58010
Ingredient               34380
Branded Dose Group       23388
Clinical Dose Group      17741
Supplier                  8593
Multiple Ingredients      4469
Precise Ingredient        3476
Branded Pack              3447
Clinical Pack             3258
Clinical Pack Box         1253
Branded Pack Box           833
Dose Form                  253
Dose Form Group             47
Name: rxnorm_concept_class_id, dtype: int64

### Codes de la bdpm dans la table des relations

In [50]:
bdpm_concept_rel = bdpm.merge(concept_rel, left_on=['bdpm_concept_id'], right_on=['concept_id_1'])
bdpm_concept_rel.head()

bdpm_code  bdpm_concept_id           bdpm_libelle bdpm_concept_class_id  \
0     00001         43047359             Loperamide            Ingredient   
1     00005         43045429   Acetylsalicylic Acid            Ingredient   
2     00007         43045800           Benzoic Acid            Ingredient   
3     00008         43045866             Boric Acid            Ingredient   
4     00011         43045611  Anhydrous Citric Acid            Ingredient   

   concept_id_1  concept_id_2 relationship_id  valid_start_date  \
0      43047359        991876         Maps to          20191006   
1      43045429       1112807         Maps to          20191006   
2      43045800        711452         Maps to          20191006   
3      43045866        929698         Maps to          20191006   
4      43045611        950435         Maps to          20191006   

   valid_end_date  invalid_reason  
0        20991231             NaN  
1        20991231             NaN  
2        20991231             NaN  
3        20991231             NaN  
4        20991231             NaN

In [51]:
# Valuers uniques par colonne
bdpm_concept_rel.nunique()

bdpm_code                43922
bdpm_concept_id          43922
bdpm_libelle             30910
bdpm_concept_class_id        6
concept_id_1             43922
concept_id_2             32256
relationship_id              3
valid_start_date             6
valid_end_date               1
invalid_reason               0
dtype: int64

#7300 / 44376 codes de la bdpm sont dans concept_relationship.

### Relations entre les codes de la BDPM et RxNorm

In [52]:
bdpm_rxnorm = bdpm_concept_rel.merge(rxnorm, left_on=['concept_id_2'], right_on=['rxnorm_concept_id']).drop(columns=['valid_start_date','valid_end_date','invalid_reason','concept_id_1','concept_id_2'])
bdpm_rxnorm.head()

bdpm_code  bdpm_concept_id                  bdpm_libelle  \
0     00001         43047359                    Loperamide   
1         1         43239732      Loperamide hydrochloride   
2         1         43239732      Loperamide hydrochloride   
3     27324         43236076  Loperamide oxide monohydrate   
4     27324         43236076  Loperamide oxide monohydrate   

  bdpm_concept_class_id     relationship_id rxnorm_code  rxnorm_concept_id  \
0            Ingredient             Maps to        6468             991876   
1            Ingredient             Maps to        6468             991876   
2            Ingredient  Source - RxNorm eq        6468             991876   
3            Ingredient             Maps to        6468             991876   
4            Ingredient  Source - RxNorm eq        6468             991876   

  rxnorm_libelle rxnorm_concept_class_id  
0     loperamide              Ingredient  
1     loperamide              Ingredient  
2     loperamide              Ingredient  
3     loperamide              Ingredient  
4     loperamide              Ingredient

In [53]:
# Nombre de valeurs uniques par colonnes
bdpm_rxnorm.nunique()

bdpm_code                  42816
bdpm_concept_id            42816
bdpm_libelle               30464
bdpm_concept_class_id          6
relationship_id                2
rxnorm_code                31150
rxnorm_concept_id          31150
rxnorm_libelle             31040
rxnorm_concept_class_id       19
dtype: int64

Seulement 6194 / 44 376 codes de la BDPM sont alignés avec un code RxNorm

In [54]:
# Nombre de valeurs uniques par classe de la BDPM
bdpm_rxnorm.bdpm_concept_class_id.value_counts()

Drug Product    21831
Brand Name      15423
Ingredient       6170
Supplier         1473
Dose Form         235
Device              2
Name: bdpm_concept_class_id, dtype: int64

In [55]:
# Nombre de valeurs unique par classe RxNorm
bdpm_rxnorm.rxnorm_concept_class_id.value_counts()

Marketed Product       21137
Brand Name             15425
Ingredient              6146
Supplier                1471
Dose Form                235
Clinical Drug            203
Quant Clinical Drug      119
Branded Drug Form        119
Clinical Drug Form       105
Branded Pack              59
Clinical Pack             28
Precise Ingredient        24
Clinical Drug Box         23
Branded Drug Box          19
Quant Branded Drug        11
Quant Branded Box          4
Branded Drug Comp          3
Branded Drug               2
Quant Clinical Box         1
Name: rxnorm_concept_class_id, dtype: int64

In [57]:
bdpm_rxnorm[bdpm_rxnorm.rxnorm_concept_class_id=='Marketed Product'].to_csv('marketed_product.csv', index=False)

In [66]:
# Nombre de valeurs uniques par type de relation
bdpm_rxnorm.relationship_id.value_counts()

Maps to               4101
Source - RxNorm eq    4022
Name: relationship_id, dtype: int64

In [42]:
# Nombre de relations 'Maps to'
bdpm_rxnorm[bdpm_rxnorm.relationship_id=='Maps to'].nunique()

bdpm_code            4101
bdpm_concept_id      4101
bdpm_libelle         3540
relationship_id         1
rxnorm_code          2232
rxnorm_concept_id    2232
rxnorm_libelle       2232
dtype: int64

In [70]:
bdpm_rxnorm.to_csv('mapping_athena_bdpm_rxnorm.csv', index=False)